### PR TITLE
Parsing Style programs

### PR DESCRIPTION
**Pull request by wodeni**
UPDATE: good to merge now.

Current status:
- Built dictionary from Substance id to style shape. For now I only deal with `Shape` commands, which is not correct. The actual type that the dictionary is mapping **to** is not clear yet. I do not think `C.StyLine` will be very appropriate because there is no one-to-one relationship from Substance objects to a `StyLine`. We will probably need a more comprehensive type for "Substance objects with style information".
- Added square class and Subset energy functions and inObj function for
Sets
- Added render functions for squares, and two other styles for dot
(commented out as if now)
- [170611] Intersection functions:
170611-intersection

- Preview:
170609-square-2
170609-square-3
170609-square

### CODE REVIEW

**Comment by wodeni** [comment]
- Support for map is now completed. We can have arrows going from/to circles/squares now!
- Previews: 
image
170614-map2

**Comment by hypotext** [comment]
Nice refactoring! Looks like the boilerplate is greatly reduced. Were you able to test all of the changed objective functions? (One thing about our codebase is that it's hard to automatically test--right now we need to manually test each program, compile into diagrams, and drag things around to verify the layout behaves correctly. One day I'd like to add automatic unit testing, perhaps on optimization positions with a set seed for the rng.)

**Comment by wodeni** [comment]
I would love to a have regression test suite in the near future. Assuming all parameters to the optimizer are the same, I think positions might just work. But the question is, what is the our ground truth here?

**Comment by wodeni** [comment]
Added a simple objective function: when a set has a subset, it should not place its own label inside of that subset.
170615-label

File: src/Runtime.hs [hunk 1]
```diff
@@ -506,84 +687,162 @@ defaultCWeight = 1
...
-                  tr "point in val: " $
-                  pointNotInExt [[xp' pt, yp' pt], [xc' set, yc' set, r' set]]
-            (_, _) -> error "point in not called on a point and a set, or 1+ doesn't exist" -- TODO check for None
-pointNotInFn _ _ = error "point in not called with 2 args"
-
-toPenalty :: (Floating a, Real a, Show a, Ord a) => (M.Map Name (Obj' a) -> a) -> (M.Map Name (Obj' a) -> a)
+
+-- Pairwise constraint functions
+
+subsetFn, noSubsetFn, intersectFn, noIntersectFn, pointInFn, pointNotInFn ::
```
Comment (hypotext) [inline src/Runtime.hs]: I would suggest splitting these up; I find that more readable

**Pull request merged**